src/osdc/Objecter.cc

```diff
@@ -3415,36 +3392,53 @@ void Objecter::list_nobjects(NListContext *list_context, Context *onfinish)
   }
   int pg_num = pool->get_pg_num();
   bool sort_bitwise = osdmap->test_flag(CEPH_OSDMAP_SORTBITWISE);
-  rl.unlock();
 
-  if (list_context->starting_pg_num == 0) {     // there can't be zero pgs!
-    list_context->starting_pg_num = pg_num;
+  if (list_context->pos.is_min()) {
+    list_context->starting_pg_num = 0;
     list_context->sort_bitwise = sort_bitwise;
-    ldout(cct, 20) << pg_num << " placement groups" << dendl;
+    list_context->starting_pg_num = pg_num;
   }
   if (list_context->sort_bitwise != sort_bitwise) {
-    ldout(cct, 10) << " hobject sort order changed, restarting this pg"
-		   << dendl;
-    list_context->cookie = collection_list_handle_t();
+    list_context->pos = hobject_t(
```

Comment:
Hmm.. possibly.  It would only matter if you set the ioctx read snapid.  The code to support this in the OSD is very slow (we have to load the object_info_t for the head and the relevant clone), so I'm wonder if we want to remove support for it altogether.

In the meantime, I guess we should just use the read snap id here?